eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs

```diff
@@ -793,30 +837,303 @@ votingDRepThresholdInternal pp isElectedCommittee action =
-freshDRepPulser n es =
-  startDRepDistr
-    n
-    (es ^. epochStateUMapL)
-    (es ^. epochStateRegDrepL)
-    (es ^. epochStateIncrStakeDistrL)
+-- Lenses for access to (DRepPulsingState era)
+
+newEpochStateDRepPulsingStateL :: ConwayEraGov era => Lens' (NewEpochState era) (DRepPulsingState era)
+newEpochStateDRepPulsingStateL = nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . drepPulsingStateGovStateL
+
+epochStateDRepPulsingStateL :: ConwayEraGov era => Lens' (EpochState era) (DRepPulsingState era)
+epochStateDRepPulsingStateL = esLStateL . lsUTxOStateL . utxosGovStateL . drepPulsingStateGovStateL
+
+-- ===============================================================================
+-- Algorithm for computing the DRep stake distrubution, with and without pulsing.
+-- ===============================================================================
+
+-- | Given three inputs
+--   1) Map (Credential 'Staking c) (DRep c).   The delegation map. Inside the DRepUView of the UMap 'um' from the DState.
+--   2) regDreps :: Map (Credential 'DRepRole c) (DRepState c). The map of registered DReps to their state. The first part of the VState.
+--   3) stakeDistr :: VMap VB VP (Credential 'Staking c) (CompactForm Coin). The aggregated stake distr extracted from the
+--      first component of the IncrementalStake i.e. (IStake credmap _) where credmap is converted to a VMap
+--  Compute the Drep distribution of stake(Coin)
+--  cost is expected to be O(size of 'stakeDistr' * log (size of 'um') * log (size of 'regDreps'))
+--  This is going to be expensive, so we will want to pulse it. Without pulsing, we estimate 3-5 seconds
+computeDrepDistr ::
+  UMap c ->
+  Map (Credential 'DRepRole c) (DRepState c) ->
+  Map (Credential 'Staking c) (CompactForm Coin) ->
+  Map (DRep c) (CompactForm Coin)
+computeDrepDistr um regDreps stakeDistr = Map.foldlWithKey' (accumDRepDistr um regDreps) Map.empty stakeDistr
+
+-- | For each 'stakecred' and coin 'c', check if that credential is delegated to some DRep.
+--   If so then add that coin to the aggregated map 'ans', mapping DReps to compact Coin
+--   If the DRep is a DRepCredential (rather than AwaysAbstain or AlwaysNoConfidence) then check
+--   that the credential is a member of the registered DRep map ('regDreps') before adding it to 'ans'
+accumDRepDistr ::
+  UMap c ->
+  Map (Credential 'DRepRole c) (DRepState c) ->
+  Map (DRep c) (CompactForm Coin) ->
+  Credential 'Staking c ->
+  CompactForm Coin ->
+  Map (DRep c) (CompactForm Coin)
+accumDRepDistr um regDreps ans stakecred c =
+  case UMap.lookup stakecred (DRepUView um) of
+    Nothing -> ans
+    Just drep@DRepAlwaysAbstain -> Map.insertWith UMap.addCompact drep c ans
+    Just drep@DRepAlwaysNoConfidence -> Map.insertWith UMap.addCompact drep c ans
+    Just drep@(DRepCredential cred2) ->
+      if Map.member cred2 regDreps
+        then Map.insertWith UMap.addCompact drep c ans
+        else ans
```

Comment:
I did this

` case UMap.lookup stakecred (DRepUView um) of
    Nothing -> ans
    Just drep@DRepAlwaysAbstain -> Map.insertWith UMap.addCompact drep c ans
    Just drep@DRepAlwaysNoConfidence -> Map.insertWith UMap.addCompact drep c ans
    Just (DRepCredential cred2) | Map.notMember cred2 regDreps -> ans
    Just (DRepCredential cred2) -> Map.insertWith UMap.addCompact drep c ans`

which has the advantage that if we ever add another constructor, it will complain.